test/common_device_type.py

```diff
@@ -135,25 +167,47 @@ def instantiated_test(self, test=test):
 
                 @wraps(test)
                 def instantiated_test(self, test=test, dtype=dtype):
-                    return test(self, cls.device_type, dtype)
+                    device_arg = cls.default_device if not hasattr(test, 'devices_required') else cls.all_devices
+                    return test(self, device_arg(), dtype)
 
                 setattr(cls, dtype_test_name, instantiated_test)
 
 
 class CPUTestBase(DeviceTypeTestBase):
-    device_type = "cpu"
+    device_type = 'cpu'
 
 
 class CUDATestBase(DeviceTypeTestBase):
-    device_type = "cuda"
+    device_type = 'cuda'
     _do_cuda_memory_leak_check = True
     _do_cuda_non_default_stream = True
 
+    @classmethod
+    def default_device(cls):
+        return 'cuda:0'
+
+    @classmethod
+    def all_devices(cls):
+        default_device_num = int(cls.default_device().split(':')[1])
```

Comment:
(I think we are in agreement about this, let me know if not)

probably worth commenting that:
1) this API isn't guaranteed to be stable
2) we are only doing it this way because we don't have a device-agnostic way of getting all the available devices for a device type

So really we shouldn't have this, it's just a chicken and egg problem.

Reply:
Correct. Actually after this PR is in let's start talking about some follow-ups in workchat.